can/io/generic.py

```diff
@@ -5,42 +5,57 @@
 """
 
 from abc import ABCMeta
+from typing import Optional, cast
+
+import can
+import can.typechecking
 
 
 class BaseIOHandler(metaclass=ABCMeta):
     """A generic file handler that can be used for reading and writing.
 
     Can be used as a context manager.
 
-    :attr file-like file:
+    :attr Optional[FileLike] file:
         the file-like object that is kept internally, or None if none
         was opened
     """
 
-    def __init__(self, file, mode="rt"):
+    def __init__(self, file: can.typechecking.AcceptedIOType, mode: str = "rt") -> None:
         """
         :param file: a path-like object to open a file, a file-like object
                      to be used as a file or `None` to not use a file at all
-        :param str mode: the mode that should be used to open the file, see
-                         :func:`open`, ignored if *file* is `None`
+        :param mode: the mode that should be used to open the file, see
+                     :func:`open`, ignored if *file* is `None`
         """
         if file is None or (hasattr(file, "read") and hasattr(file, "write")):
             # file is None or some file-like object
-            self.file = file
+            self.file = cast(Optional[can.typechecking.FileLike], file)
```

Comment:
Can we can maybe refactor this to use `isinstance` instead and avoid the explicit `cast`? A series of `if...elif...` `isinstance` checks is completely type-checked, while a `cast` simply silences the type-checker.

Reply:
I didn't want to do that since the point of a file-*like* object is that it may reside in memory or whatever, and thus not inherit from the typical file class. Keeping duck-typing alive in this case seems more important for me. But if you can do that with `isinstance` as well somehow, we can surely change it.